ports/stm32f4/common-hal/neopixel_write/__init__.c

```diff
@@ -63,23 +68,26 @@ void common_hal_neopixel_write (const digitalio_digitalinout_obj_t* digitalinout
 
     __disable_irq();
     // Enable DWT in debug core. Useable when interrupts disabled, as opposed to Systick->VAL
-    //ITM->LAR = 0xC5ACCE55; //this should be required but isn't
     CoreDebug->DEMCR |= CoreDebug_DEMCR_TRCENA_Msk;
     DWT->CTRL |= DWT_CTRL_CYCCNTENA_Msk;
     DWT->CYCCNT = 0;
 
     for(;;) {
+        cyc = (pix & mask) ? t1 : t0;
         start = DWT->CYCCNT;
         LL_GPIO_SetOutputPin(p_port, p_mask);
-        cyc = (pix & mask) ? t1 : t0;
-        while(DWT->CYCCNT - start < cyc);
+        while((DWT->CYCCNT - start) < cyc) {
+            __asm__ __volatile__("nop");
```

Comment:
Why have nop here? Doesn't it just make your single loop iteration slower and therefore less granular?

Reply:
Not having it here would cause the compiler to optimize the loop out under certain settings. For a while having Os optimization would cause the entire board to hang, and I never quite figured that out. But I can revisit removing these if you think it'd provide benefits.

Reply:
Isn't `DWT->CYCCNT` a volatile read? It shouldn't be optimized out then.

Reply:
That's a good point. I'm not sure what the optimizer was doing in that case then, it was definitely causing some weird timing when I was varying between different settings. But testing the latest version with them gone seems to work fine, so I'll remove them.